File: examples/android/TextToSpeech/app/src/main/java/ai/moonshine/examples/texttospeech/TtsAssetDownloader.kt
package ai.moonshine.examples.texttospeech

import ai.moonshine.voice.TextToSpeech
import ai.moonshine.voice.TranscriberOption
import java.io.File
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import org.json.JSONArray

/**
 * Resolves TTS asset dependencies via `moonshine_get_tts_dependencies` and downloads any missing
 * files from the Moonshine CDN (`https://download.moonshine.ai/tts/<relative/key>`) into
 * `<g2pRoot>/<relative/key>`.
 */
object TtsAssetDownloader {

    private const val CDN_BASE = "https://download.moonshine.ai/tts/"
    private const val TMP_SUFFIX = ".download"

    fun interface ProgressListener {
        /**
         * @param key current asset key (e.g. `en_us/dict.tsv`).
         * @param fileIndex 1-based index of the file being downloaded.
         * @param totalFiles total number of files that will be downloaded this pass.
         * @param bytesDownloaded bytes written for this file so far.
         * @param bytesTotal total bytes for this file (may be `-1` if the server did not report `Content-Length`).
         */
        fun onProgress(
            key: String,
            fileIndex: Int,
            totalFiles: Int,
            bytesDownloaded: Long,
            bytesTotal: Long,
        )
    }

    /**
     * Make sure every asset required for `language` + optional prefixed `voice` (e.g.
     * `kokoro_af_alloy`, `piper_en_US-ryan-low`) exists under `g2pRoot`. Downloads each missing
     * file synchronously; callers must invoke this off the main thread.
     */
    fun ensureAssetsPresent(
        g2pRoot: File,
        language: String,
        voice: String?,
        listener: ProgressListener,
    ) {
        val keys = resolveDependencyKeys(g2pRoot, language, voice)
        val missing =
            keys.filter { key ->
                key.isNotEmpty() && key.contains('/') && !File(g2pRoot, key).exists()
            }
        if (missing.isEmpty()) return

        missing.forEachIndexed { index, key ->
            val dest = File(g2pRoot, key)
            dest.parentFile?.mkdirs()
            downloadOne(key, dest, index + 1, missing.size, listener)
        }
    }

    private fun resolveDependencyKeys(
        g2pRoot: File,
        language: String,
        voice: String?,
    ): List<String> {
        val opts = ArrayList<TranscriberOption>()
        opts.add(TranscriberOption("g2p_root", g2pRoot.absolutePath))
        if (voice != null) {
            opts.add(TranscriberOption("voice", voice))
        }
        val json = TextToSpeech.getTtsDependencies(language, opts)
        val arr = JSONArray(json)
        val out = ArrayList<String>(arr.length())
        for (i in 0 until arr.length()) {
            out.add(arr.optString(i, ""))
        }
        return out
    }

    private fun downloadOne(
        key: String,
        dest: File,
        fileIndex: Int,
        totalFiles: Int,
        listener: ProgressListener,
    ) {
        val url = URL(CDN_BASE + encodeKey(key))
        val tmp = File(dest.parentFile, dest.name + TMP_SUFFIX)
        tmp.delete()

        val conn = url.openConnection() as HttpURLConnection
        conn.connectTimeout = 15_000
        conn.readTimeout = 60_000
        conn.instanceFollowRedirects = true
        try {
            conn.connect()
            val code = conn.responseCode
            if (code !in 200..299) {
                throw IOException("HTTP $code fetching $url")
            }
            val totalBytes = conn.contentLengthLong
            listener.onProgress(key, fileIndex, totalFiles, 0L, totalBytes)

            conn.inputStream.use { input ->
                tmp.outputStream().use { output ->
                    val buffer = ByteArray(64 * 1024)
                    var downloaded = 0L
                    var lastReported = 0L
                    while (true) {
                        val read = input.read(buffer)
                        if (read < 0) break
                        output.write(buffer, 0, read)
                        downloaded += read
                        if (downloaded - lastReported >= 64 * 1024 || read == 0) {
                            listener.onProgress(key, fileIndex, totalFiles, downloaded, totalBytes)
                            lastReported = downloaded
                        }
                    }
                    listener.onProgress(key, fileIndex, totalFiles, downloaded, totalBytes)
                }
            }
        } finally {
            conn.disconnect()
        }

        if (!tmp.renameTo(dest)) {
            tmp.delete()
            throw IOException("Failed to move download into place: ${dest.absolutePath}")
        }
    }

    /** Percent-encode each path segment; the library expects canonical `/`-joined relative keys. */
    private fun encodeKey(key: String): String =
        key.split('/').joinToString("/") { segment ->
            java.net.URLEncoder.encode(segment, "UTF-8").replace("+", "%20")
        }
}

File: examples/ios/TextToSpeech/TextToSpeech/TtsAssetDownloader.swift
import Foundation
import MoonshineVoice

/// Resolves TTS asset dependencies via `TextToSpeech.getDependencies` and downloads any missing
/// files from the Moonshine CDN (`https://download.moonshine.ai/tts/<relative/key>`) into
/// `<g2pRoot>/<relative/key>`.
enum TtsAssetDownloader {
    private static let cdnBase = "https://download.moonshine.ai/tts/"
    private static let tmpSuffix = ".download"

    struct Progress {
        /// Relative asset key being fetched (e.g. `en_us/dict.tsv`).
        let key: String
        /// 1-based index of the file being downloaded this pass.
        let fileIndex: Int
        /// Total number of files that will be downloaded this pass.
        let totalFiles: Int
        /// Bytes written for the current file so far.
        let bytesDownloaded: Int64
        /// Total bytes for the current file (`-1` if the server did not report `Content-Length`).
        let bytesTotal: Int64
    }

    enum DownloadError: LocalizedError {
        case httpStatus(Int, URL)
        case badResponse(URL)
        case move(URL, Error)

        var errorDescription: String? {
            switch self {
            case .httpStatus(let code, let url):
                return "HTTP \(code) fetching \(url.lastPathComponent)"
            case .badResponse(let url):
                return "Invalid response for \(url.lastPathComponent)"
            case .move(let url, let error):
                return "Failed to save \(url.lastPathComponent): \(error.localizedDescription)"
            }
        }
    }

    /// Make sure every asset required for `language` + optional prefixed `voice`
    /// (e.g. `kokoro_af_alloy`, `piper_en_US-ryan-low`) exists under `g2pRoot`.
    /// Downloads each missing file synchronously; callers must invoke this off the main actor.
    static func ensureAssetsPresent(
        g2pRoot: URL,
        language: String,
        voice: String?,
        onProgress: @Sendable @escaping (Progress) -> Void
    ) async throws {
        let keys = try resolveDependencyKeys(
            g2pRoot: g2pRoot, language: language, voice: voice)
        let missing = keys.filter { key in
            !key.isEmpty && key.contains("/")
                && !FileManager.default.fileExists(
                    atPath: g2pRoot.appendingPathComponent(key).path)
        }
        if missing.isEmpty { return }

        for (index, key) in missing.enumerated() {
            try await downloadOne(
                key: key,
                dest: g2pRoot.appendingPathComponent(key),
                fileIndex: index + 1,
                totalFiles: missing.count,
                onProgress: onProgress
            )
        }
    }

    private static func resolveDependencyKeys(
        g2pRoot: URL, language: String, voice: String?
    ) throws -> [String] {
        var options: [TranscriberOption] = [
            TranscriberOption(name: "g2p_root", value: g2pRoot.path)
        ]
        if let voice = voice {
            options.append(TranscriberOption(name: "voice", value: voice))
        }
        let json = try TextToSpeech.getDependencies(
            languages: language, options: options)
        guard let data = json.data(using: .utf8),
            let array = try JSONSerialization.jsonObject(with: data) as? [String]
        else { return [] }
        return array
    }

    private static func downloadOne(
        key: String,
        dest: URL,
        fileIndex: Int,
        totalFiles: Int,
        onProgress: @Sendable @escaping (Progress) -> Void
    ) async throws {
        guard let url = cdnURL(for: key) else {
            throw URLError(.badURL)
        }

        try FileManager.default.createDirectory(
            at: dest.deletingLastPathComponent(),
            withIntermediateDirectories: true)

        let tmp = dest.appendingPathExtension(String(tmpSuffix.dropFirst()))
        try? FileManager.default.removeItem(at: tmp)

        let (bytes, response) = try await URLSession.shared.bytes(from: url)
        guard let http = response as? HTTPURLResponse else {
            throw DownloadError.badResponse(url)
        }
        guard (200...299).contains(http.statusCode) else {
            throw DownloadError.httpStatus(http.statusCode, url)
        }
        let totalBytes = http.expectedContentLength

        onProgress(
            Progress(
                key: key, fileIndex: fileIndex, totalFiles: totalFiles,
                bytesDownloaded: 0, bytesTotal: totalBytes))

        FileManager.default.createFile(atPath: tmp.path, contents: nil)
        guard let handle = try? FileHandle(forWritingTo: tmp) else {
            throw DownloadError.badResponse(url)
        }
        defer { try? handle.close() }

        var buffer = Data()
        buffer.reserveCapacity(64 * 1024)
        var downloaded: Int64 = 0
        var lastReported: Int64 = 0

        for try await byte in bytes {
            buffer.append(byte)
            if buffer.count >= 64 * 1024 {
                try handle.write(contentsOf: buffer)
                downloaded += Int64(buffer.count)
                buffer.removeAll(keepingCapacity: true)
                if downloaded - lastReported >= 64 * 1024 {
                    onProgress(
                        Progress(
                            key: key, fileIndex: fileIndex, totalFiles: totalFiles,
                            bytesDownloaded: downloaded, bytesTotal: totalBytes))
                    lastReported = downloaded
                }
            }
        }
        if !buffer.isEmpty {
            try handle.write(contentsOf: buffer)
            downloaded += Int64(buffer.count)
        }
        try handle.close()

        onProgress(
            Progress(
                key: key, fileIndex: fileIndex, totalFiles: totalFiles,
                bytesDownloaded: downloaded, bytesTotal: totalBytes))

        if FileManager.default.fileExists(atPath: dest.path) {
            try FileManager.default.removeItem(at: dest)
        }
        do {
            try FileManager.default.moveItem(at: tmp, to: dest)
        } catch {
            try? FileManager.default.removeItem(at: tmp)
            throw DownloadError.move(dest, error)
        }
    }

    /// Percent-encode each path segment; the library expects canonical `/`-joined relative keys.
    private static func cdnURL(for key: String) -> URL? {
        let allowed = CharacterSet.urlPathAllowed.subtracting(CharacterSet(charactersIn: "/"))
        let encoded =
            key
            .split(separator: "/", omittingEmptySubsequences: false)
            .map { segment -> String in
                String(segment).addingPercentEncoding(withAllowedCharacters: allowed) ?? String(
                    segment)
            }
            .joined(separator: "/")
        return URL(string: cdnBase + encoded)
    }
}
